src/main/scala/com/databricks/spark/csv/CsvRelation.scala

```diff
@@ -15,32 +15,34 @@
  */
 package com.databricks.spark.csv
 
-import org.apache.hadoop.fs.{FileSystem, Path}
-import org.slf4j.LoggerFactory
+import java.io.IOException
 
-import org.apache.commons.csv._
+import scala.collection.JavaConversions._
+import scala.util.control.NonFatal
 
+import org.apache.commons.csv._
+import org.apache.hadoop.fs.Path
 import org.apache.spark.sql._
 import org.apache.spark.sql.catalyst.expressions._
-import org.apache.spark.sql.sources.TableScan
+import org.apache.spark.sql.sources.{BaseRelation, InsertableRelation, TableScan}
 import org.apache.spark.sql.types.{StructType, StructField, StringType}
+import org.slf4j.LoggerFactory
 
-import scala.collection.JavaConversions._
-import scala.util.control.NonFatal
 
 case class CsvRelation protected[spark] (
     location: String,
     useHeader: Boolean,
     delimiter: Char,
     quote: Char,
-    userSchema: StructType = null)(@transient val sqlContext: SQLContext) extends TableScan {
+    userSchema: StructType = null)(@transient val sqlContext: SQLContext)
+  extends BaseRelation with TableScan with InsertableRelation {
 
   private val logger = LoggerFactory.getLogger(CsvRelation.getClass)
 
-  val schema = inferSchema()
+  var schema = inferSchema()
 
   // By making this a lazy val we keep the RDD around, amortizing the cost of locating splits.
-  lazy val buildScan = {
+  def buildScan = {
```

Comment:
If you change `lazy val` to `def`, `INSERT OVERWRITE` works. Otherwise the problem that I brought up occurs.